### PR TITLE
integration Specs

### PR DESCRIPTION
**Pull request by shahzaman900**
### Implemented Changes:
- Using the where method the posts data is being fetched and the comments are loaded for each post causing n+1 problem.
- Updated the index.html.erb of posts, where the iteration of five_recent_comments using the includes eager loading method.
- Created a new folder views in spec for implementing integration tests for the views using the Capybara gem.
- Added test cases that checks the view working from the Capybara view screenshots from the created mock data.

281246618-0eb77815-3e19-4573-93ac-c67c77f01401
281261247-588f70ac-cf01-4aad-af97-7ea4dc88867b

### CODE REVIEW

File: spec/features/post_show_spec.rb [hunk 1]
```diff
@@ -1,19 +1,19 @@
...
-    3.times { |i| Post.create(author: @fu, title: "Post #{i + 1}", text: "First author Post #{i + 1} body") }
-    5.times { |i| Post.create(author: @su, title: "Post #{i + 1}", text: "Second author Post #{i + 1} body") }
+    3.times { |i| Post.created!(author: @fu, title: "Post #{i + 1}", text: "First author Post #{i + 1} body") }
+    5.times { |i| Post.created!(author: @su, title: "Post #{i + 1}", text: "Second author Post #{i + 1} body") }
 
-    6.times { |i| Comment.create(post: @fu.posts.first, user: @fu, text: "Comment #{i + 1}") }
+    6.times { |i| Comment.created!(post: @fu.posts.first, user: @fu, text: "Comment #{i + 1}") }
     3.times { |i| Comment.create(post: @su.posts.first, user: @su, text: "Comment #{i + 1}") }
 
     6.times { Like.create(post: @fu.posts.first, user: @fu) }
```
Comment (Whoistolu) [inline spec/features/post_show_spec.rb]: - [ ] [REQUIRED] Kindly add a test that asserts you can see the username of each commentor.
- [ ] [REQUIRED] Kindly add tests that assert you can see the comment each commentor left.
These are hard requirements, let's fix this team.

File: app/views/posts/index.html.erb [hunk 2]
```diff
@@ -14,9 +14,9 @@
         </div>
     <% end -%>
 
-    <% if (current_user.id == @user.id) && (can? :create, Post) %>
-      <h2 class="add-post"><%= link_to 'Add new Post', new_user_post_path %></h2>
-    <% end %>
+    <%# <% if (current_user.id == @user.id) && (can? :create, Post) %>
+      <%# <h2 class="add-post"><%= link_to 'Add new Post', new_user_post_path</h2>  %>
+    <%# <% end %>
```
Comment (Whoistolu) [inline app/views/posts/index.html.erb]: - [ ] [OPTIONAL] Kindly remove all unused comments

**Review by Whoistolu — CHANGES_REQUESTED** [review]
Hi @sheezu900 , @Graciano1997 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Required Changes ♻️

- [x]  _Check the inline comments under the review for sections to improve._
- [x] NB: I am adding this recommendation here because they are not present in the file changes

- [x] [REQUIRED] There are some test missing for the featured tests for `User` show page
- Tests that redirects to the post's show page when you click on a user's post.
- Tests that redirects to the user's post's index page when you click on a **see all posts**.

- [x] [REQUIRED] There are some test missing in the featured tests for `User` **post index page**
- Kindly add tests that asserts you can see a post's title.
- Tests that asserts you can see the first comments on a post.
- Tests that redirects to the post's show page when you click on a post.

### Optional suggestions
Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better.

Cheers and Happy coding!

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me @whoistolu in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/features/user_show_spec.rb [hunk 3]
```diff
@@ -100,5 +100,26 @@
...
+    end
+
+    it 'should redirect to users posts index page when you click on a see all posts.' do
+      visit posts_path
+      sleep(1)
+      click_link('See all posts')
+      sleep(1)
+      expect(current_path).to eq(user_posts_path(@su))
+    end
+    end
```
Comment (Gambit142) [inline spec/features/user_show_spec.rb]: - [x] There is a syntax error here and it is affecting the tests from running. Kindly remove one `end` statement here. 👍 😎

File: spec/features/post_show_spec.rb [hunk 4]
```diff
@@ -1,6 +1,6 @@
 require 'rails_helper'
 
-RSpec.describe 'Posts', type: :system do
+RSpec.describe 'Posts', type: :feature do
   before :all do
     Like.delete_all
     Comment.delete_all
```
Comment (Gambit142) [inline spec/features/post_show_spec.rb:6]: ### Although I am highlighting this file, this suggested change should be done in the `spec/features/posts_index_spec.rb` file. 

- [ ] As highlighted by the previous reviewer, kindly do the following:

- Write a test that redirects to the `Posts#Show` page when you click on a post on the `Posts#Index` page. 

The code you wrote in the screenshot below in the `spec/features/posts_index_spec.rb` file is directing to the `Posts#Show` page from the `Users#Show` page and not the `Posts#Index` page. Kindly write the correct test 👍 :

image

File: spec/features/user_show_spec.rb (hunk 3 above)
Comment (Gambit142) [inline spec/features/user_show_spec.rb:119]: - [ ] After you implemented some changes that the first reviewer suggested, I noticed that 30 tests out of 34 tests are failing. Kindly fix the tests and ensure that all tests pass after implementing all changes. 💻 😎 

image

**Review by Gambit142 — CHANGES_REQUESTED** [review]
Hi @sheezu900 @Graciano1997 👋, 

#### Great work on making the changes requested by a previous reviewer and fixing the error that made most tests not pass 👏 

### Highlights
- Setup linters correctly ✔️ 
- Use of Rsec testing library ✔️ 
- Wrote feature spec tests ✔️

You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

 ## Required Changes ♻️
_Kindly check the comments under the review._

## Optional suggestions

N/A

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you consider them as they can improve your code. Some of them were simply missed by the previous reviewer and addressing them will improve your application._


Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

#### Please tag me in your question so I can receive the notification. You can tag me by typing @Gambit142

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/features/user_show_spec.rb [hunk 5]
```diff
@@ -101,4 +101,4 @@
       expect(current_path).to eq(user_posts_path(@su))
     end
   end
-end
\ No newline at end of file
+encouraged
```
Comment (codepantha) [inline spec/features/user_show_spec.rb]: - [ ] This line here prevents the code from running correctly as can be seen in the attached image below 👇 
Kindly remove `encouraged` and replace it with `end` 👍 

image

**Review by codepantha — CHANGES_REQUESTED** [review]
# STATUS: CHANGES REQUIRED ♻️ 

## Hi team 👋, 

Great work on making the changes 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.



## Highlights 🎯 
- Setup linters correctly ✔️
- Use of Rsec testing library ✔️
- Wrote feature spec tests ✔️


## Suggested changes
_Check the comments under the review._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

**Review by Shedrack-Sunday — APPROVED** [review]
Hi @sheezu900       👍 

Your project is complete! There is nothing else to say other than... it's time to merge it :shipit:
Congratulations! 🎉

image


Highlights
- [x] Good project organization ✔️
- [x] Nice use of Rails  features  ✔️  
- [x] Kudus on meeting the requirements as pointed the last time out   👏 


Optional suggestions
Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application.


Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s).

**Pull request merged**